src/chef_sql.erl

```diff
@@ -544,6 +544,54 @@ fetch_environment_filtered_recipes(OrgId, Environment) ->
             {error, Reason}
     end.
 
+-spec bulk_fetch_cookbook_versions(OrgId::object_id(), [versioned_cookbook()]) ->
+                                          [#chef_cookbook_version{}].
+bulk_fetch_cookbook_versions(OrgId, CookbookVersions) ->
+    QueryParam = cookbook_versions_array_to_binary(CookbookVersions),
+    case sqerl:select(bulk_fetch_cookbook_versions, [OrgId, QueryParam], ?ALL(chef_cookbook_version)) of
+        {ok, none} ->
+            [];
+        {ok, Results} ->
+            Results;
+        {error, Error} ->
+            {error, Error}
+    end.
+
+-spec cookbook_versions_array_to_binary([versioned_cookbook()]) -> binary().
+cookbook_versions_array_to_binary(CookbookVersions) ->
+    cookbook_versions_array_to_binary(CookbookVersions, <<"{">>, <<"}">>, <<"">>).
+
+-spec cookbook_versions_array_to_binary([versioned_cookbook()], binary(), binary(), binary()) -> binary().
+cookbook_versions_array_to_binary([CkbVer|CookbookVersions], Acc, EndBin, Sep) ->
+    CkbBin = cookbook_version_to_binary(CkbVer),
+    cookbook_versions_array_to_binary(CookbookVersions,
+                                      <<Acc/binary, Sep/binary, CkbBin/binary>>,
+                                      EndBin, <<",">>);
+cookbook_versions_array_to_binary([], Acc, EndBin, _Sep) ->
+    <<Acc/binary, EndBin/binary>>.
+
+
+%% @doc Tranform a versioned_cookbook() into a binary that can be used
+%% in the bulk_fetch_cookbook_version sql query.  A
+%% versioned_cookbook() looks like:
+%%   {binary(), {integer(), integer(), integer()}}
+%%
+%% for example:
+%%  {<<"yum">>, {0, 2, 43}}
+%%
+%% our desired output for this example would be:
+%%   <<"\"(yum, 0, 2, 43)\"">>
+%%
+%%
+-spec cookbook_version_to_binary(versioned_cookbook()) -> binary().
+cookbook_version_to_binary(CkbVer) ->
+    Name = element(1, CkbVer),
+    Version = element(2, CkbVer),
+    Major = list_to_binary(integer_to_list(element(1, Version))),
+    Minor = list_to_binary(integer_to_list(element(2, Version))),
+    Patch = list_to_binary(integer_to_list(element(3, Version))),
+    <<"\"(", Name/binary, ",", Major/binary, ",", Minor/binary, ",", Patch/binary, ")\"">>.
```

Comment:
We can avoid the separate assignments and intermediate int->list->binary calls with: 

    iolist_to_binary([<<"\"(">>, Name, <<",">>, 
                     integer_to_binary(Major), <<",">>, 
                     integer_to_binary(Minor), <<",">>, 
                     integer_to_binary(Patch), <<")\"">>]).

`integer_to_binary`  is R16+ only, but that's safe since we're on R16B3 and hopefully soon to move to 18.

iirc iolist_to_binary moves the heavy lifting out of the VM and into a native function, but I'll run some comparisons locally and update here.